python/beeai_framework/adapters/litellm/embedding.py

```diff
@@ -0,0 +1,134 @@
+# limitations under the License.
+
+
+import logging
+import os
+from abc import ABC
+from itertools import chain
+from typing import Any
+
+import litellm
+from litellm import aembedding
+from litellm.litellm_core_utils.get_supported_openai_params import get_supported_openai_params
+from litellm.types.utils import EmbeddingResponse
+
+from beeai_framework.backend import EmbeddingModel
+from beeai_framework.backend.types import EmbeddingModelInput, EmbeddingModelOutput, EmbeddingModelUsage
+from beeai_framework.context import RunContext
+from beeai_framework.logger import Logger
+
+logger = Logger(__name__)
+
+
+class LiteLLMEmbeddingModel(EmbeddingModel, ABC):
+    def __init__(
+        self,
+        model_id: str,
+        *,
+        provider_id: str,
+        **kwargs: dict[str, Any],
+    ) -> None:
+        self._settings: dict[str, Any] = kwargs
+        self._model_id = model_id
+        self._litellm_provider_id = provider_id
+        self.supported_params = get_supported_openai_params(model=self.model_id, custom_llm_provider=provider_id) or []
+        # drop any unsupported parameters that were passed in
+        litellm.drop_params = True
+        # disable LiteLLM caching in favor of our own
+        litellm.disable_cache()  # type: ignore [attr-defined]
+
+    @property
+    def model_id(self) -> str:
+        return self._model_id
+
+    @staticmethod
+    def litellm_debug(enable: bool = True) -> None:
+        litellm.set_verbose = enable  # type: ignore [attr-defined]
+        litellm.suppress_debug_info = not enable
+
+        litellm.suppress_debug_info = not enable
+        litellm.logging = enable
+
+        logger = logging.getLogger("LiteLLM")
+        logger.setLevel(logging.DEBUG if enable else logging.CRITICAL + 1)
+
+    async def _create(
+        self,
+        input: EmbeddingModelInput,
+        run: RunContext,
+    ) -> EmbeddingModelOutput:
+        litellm_input = self._transform_input(input)
+        response = await aembedding(**litellm_input)
+        response_output = self._transform_output(response, input)
+        logger.debug(f"Inference response output:\n{response_output}")
+        return response_output
+
+    def _transform_input(self, model_input: EmbeddingModelInput) -> dict[str, Any]:
+        return {
+            "model": self.provider_id + "/" + self._model_id,
+            "input": model_input.values,
+            **self._settings,
+        }
+
+    def _transform_output(self, response: EmbeddingResponse, model_input: EmbeddingModelInput) -> EmbeddingModelOutput:
+        embeddings: list[list[float]] = []
+
+        for embedding in response.data:
+            embeddings.append([float(point) for point in embedding.get("embedding")])
+
+        return EmbeddingModelOutput(
+            values=model_input.values,
+            embeddings=embeddings,
+            usage=EmbeddingModelUsage(**response.usage.model_dump()) if response.usage else None,
```

Comment:
I would extend the `EmbeddingModelOutput` to include a raw response (untyped obviously).